service/qos/service_level_controller.cc

```diff
@@ -230,16 +230,41 @@ future<std::optional<service_level_options>> service_level_controller::find_serv
     });
 }
 
-future<>  service_level_controller::notify_service_level_added(sstring name, service_level sl_data) {
-    _service_levels_db.emplace(name, sl_data);
-    return make_ready_future();
+future<>  service_level_controller::notify_service_level_added(sstring name, service_level sl_data) {    
+    return parallel_for_each(_subscribers, [name, sl_data] (subscriber_registry_entry& e) {
+        if (e->_qccs) {
+            return e->_qccs->on_before_service_level_add(name, sl_data.slo);
+        } else {            
+            return make_ready_future();
+        }
+    }).then([this, name, sl_data] () {
+        std::remove_if(_subscribers.begin(), _subscribers.end(), [] (subscriber_registry_entry& e) {
+            return *e == (qos_configuration_change_subscriber*)nullptr;
+        });
+        _service_levels_db.emplace(name, sl_data);
+        return make_ready_future();
+    });
+    
 }
 
 future<> service_level_controller::notify_service_level_updated(sstring name, service_level_options slo) {
     auto sl_it = _service_levels_db.find(name);
     future<> f = make_ready_future();
     if (sl_it != _service_levels_db.end()) {
-        sl_it->second.slo = slo;
+        service_level_options slo_before = sl_it->second.slo;
+        return parallel_for_each(_subscribers, [name, slo_before, slo] (subscriber_registry_entry& e) {
+            if (e->_qccs) {
+                return e->_qccs->on_before_service_level_change(name, slo_before, slo);
+            } else {
+                return make_ready_future();
+            }
+        }).then([this, sl_it, slo] () {
+            std::remove_if(_subscribers.begin(), _subscribers.end(), [] (subscriber_registry_entry& e) {
+                return *e == (qos_configuration_change_subscriber*)nullptr;
+            });
```

Comment:
I don't understand this remove_if. First, std::remove_if() doesn't remove anything, it just rearranges the input. Second, why remove subscribers on service level updates?

Reply:
It is removed in the new version, I tried to reinvent the wheel for protection from deregistration. It was a bad idea...